third_party/tsl/third_party/gpus/cuda/hermetic/cuda_redist_init_repositories.bzl

```diff
@@ -472,27 +473,55 @@ def cudnn_redist_init_repository(
         cudnn_redist_path_prefix = cudnn_redist_path_prefix,
     )
 
+def detect_platform(repository_ctx):
+    """Detect if the platform is x86_64, Jetson (aarch64), or generic ARM SBSA."""
+    # Use environment variables to distinguish between platforms
+    host_arch = repository_ctx.os.arch  # Using Bazel's built-in context to detect architecture
+    
+    if host_arch == "x86_64":
+        return "linux-x86_64"  # x86_64 platform
+    
+    if host_arch == "aarch64":
+        # Check if the build is targeting Jetson using an environment variable
+        is_jetson = repository_ctx.os.environ.get("JETSON_PLATFORM", None)
```

Comment:
I will try. Sorry for the delay

Reply:
> Instead of using environment variables, we can utilize the output of [uname](https://ss64.com/bash/uname.html) command.
> 
> What does it return for Jetson platform?

johnny@johnny-jetson:~/Projects$ uname -a
Linux johnny-jetson 5.15.136-tegra #1 SMP PREEMPT Mon May 6 09:56:39 PDT 2024 aarch64 aarch64 aarch64 GNU/Linux
johnny@johnny-jetson:~/Projects$ uname -p
aarch64
johnny@johnny-jetson:~/Projects$ uname -i
aarch64

Reply:
Looks like we can use the output of the command `uname -a` to distinguish Tegra devices from others.
I'll try to create a PR shortly and will give you the link.

Reply:
Feel free

Reply:
```
WARNING: The following configs were expanded more than once: [cuda]. For repeatable flags, repeats are counted twice and may lead to unexpected behavior.
INFO: Repository cuda_nvjitlink instantiated at:
  /home/johnny/Projects/jax/WORKSPACE:81:30: in <toplevel>
  /home/johnny/.cache/bazel/_bazel_johnny/ec3d842aa30ca8685e26a318bc6b64a6/external/tsl/third_party/gpus/cuda/hermetic/cuda_redist_init_repositories.bzl:538:18: in cuda_redist_init_repositories
Repository rule cuda_repo defined at:
  /home/johnny/.cache/bazel/_bazel_johnny/ec3d842aa30ca8685e26a318bc6b64a6/external/tsl/third_party/gpus/cuda/hermetic/cuda_redist_init_repositories.bzl:350:28: in <toplevel>
INFO: Repository cuda_cupti instantiated at:
  /home/johnny/Projects/jax/WORKSPACE:81:30: in <toplevel>
  /home/johnny/.cache/bazel/_bazel_johnny/ec3d842aa30ca8685e26a318bc6b64a6/external/tsl/third_party/gpus/cuda/hermetic/cuda_redist_init_repositories.bzl:538:18: in cuda_redist_init_repositories
Repository rule cuda_repo defined at:
  /home/johnny/.cache/bazel/_bazel_johnny/ec3d842aa30ca8685e26a318bc6b64a6/external/tsl/third_party/gpus/cuda/hermetic/cuda_redist_init_repositories.bzl:350:28: in <toplevel>
INFO: Repository cuda_nvcc instantiated at:
  /home/johnny/Projects/jax/WORKSPACE:81:30: in <toplevel>
  /home/johnny/.cache/bazel/_bazel_johnny/ec3d842aa30ca8685e26a318bc6b64a6/external/tsl/third_party/gpus/cuda/hermetic/cuda_redist_init_repositories.bzl:538:18: in cuda_redist_init_repositories
Repository rule cuda_repo defined at:
  /home/johnny/.cache/bazel/_bazel_johnny/ec3d842aa30ca8685e26a318bc6b64a6/external/tsl/third_party/gpus/cuda/hermetic/cuda_redist_init_repositories.bzl:350:28: in <toplevel>
INFO: Repository cuda_curand instantiated at:
  /home/johnny/Projects/jax/WORKSPACE:81:30: in <toplevel>
  /home/johnny/.cache/bazel/_bazel_johnny/ec3d842aa30ca8685e26a318bc6b64a6/external/tsl/third_party/gpus/cuda/hermetic/cuda_redist_init_repositories.bzl:538:18: in cuda_redist_init_repositories
Repository rule cuda_repo defined at:
  /home/johnny/.cache/bazel/_bazel_johnny/ec3d842aa30ca8685e26a318bc6b64a6/external/tsl/third_party/gpus/cuda/hermetic/cuda_redist_init_repositories.bzl:350:28: in <toplevel>
INFO: Repository cuda_cudnn instantiated at:
  /home/johnny/Projects/jax/WORKSPACE:85:29: in <toplevel>
  /home/johnny/.cache/bazel/_bazel_johnny/ec3d842aa30ca8685e26a318bc6b64a6/external/tsl/third_party/gpus/cuda/hermetic/cuda_redist_init_repositories.bzl:513:15: in cudnn_redist_init_repository
Repository rule cudnn_repo defined at:
  /home/johnny/.cache/bazel/_bazel_johnny/ec3d842aa30ca8685e26a318bc6b64a6/external/tsl/third_party/gpus/cuda/hermetic/cuda_redist_init_repositories.bzl:433:29: in <toplevel>
INFO: Repository cuda_cusolver instantiated at:
  /home/johnny/Projects/jax/WORKSPACE:81:30: in <toplevel>
  /home/johnny/.cache/bazel/_bazel_johnny/ec3d842aa30ca8685e26a318bc6b64a6/external/tsl/third_party/gpus/cuda/hermetic/cuda_redist_init_repositories.bzl:538:18: in cuda_redist_init_repositories
Repository rule cuda_repo defined at:
  /home/johnny/.cache/bazel/_bazel_johnny/ec3d842aa30ca8685e26a318bc6b64a6/external/tsl/third_party/gpus/cuda/hermetic/cuda_redist_init_repositories.bzl:350:28: in <toplevel>
INFO: Repository cuda_nvml instantiated at:
  /home/johnny/Projects/jax/WORKSPACE:81:30: in <toplevel>
  /home/johnny/.cache/bazel/_bazel_johnny/ec3d842aa30ca8685e26a318bc6b64a6/external/tsl/third_party/gpus/cuda/hermetic/cuda_redist_init_repositories.bzl:538:18: in cuda_redist_init_repositories
Repository rule cuda_repo defined at:
  /home/johnny/.cache/bazel/_bazel_johnny/ec3d842aa30ca8685e26a318bc6b64a6/external/tsl/third_party/gpus/cuda/hermetic/cuda_redist_init_repositories.bzl:350:28: in <toplevel>
INFO: Repository cuda_cufft instantiated at:
  /home/johnny/Projects/jax/WORKSPACE:81:30: in <toplevel>
  /home/johnny/.cache/bazel/_bazel_johnny/ec3d842aa30ca8685e26a318bc6b64a6/external/tsl/third_party/gpus/cuda/hermetic/cuda_redist_init_repositories.bzl:538:18: in cuda_redist_init_repositories
Repository rule cuda_repo defined at:
  /home/johnny/.cache/bazel/_bazel_johnny/ec3d842aa30ca8685e26a318bc6b64a6/external/tsl/third_party/gpus/cuda/hermetic/cuda_redist_init_repositories.bzl:350:28: in <toplevel>
INFO: Repository cuda_cublas instantiated at:
  /home/johnny/Projects/jax/WORKSPACE:81:30: in <toplevel>
  /home/johnny/.cache/bazel/_bazel_johnny/ec3d842aa30ca8685e26a318bc6b64a6/external/tsl/third_party/gpus/cuda/hermetic/cuda_redist_init_repositories.bzl:538:18: in cuda_redist_init_repositories
Repository rule cuda_repo defined at:
  /home/johnny/.cache/bazel/_bazel_johnny/ec3d842aa30ca8685e26a318bc6b64a6/external/tsl/third_party/gpus/cuda/hermetic/cuda_redist_init_repositories.bzl:350:28: in <toplevel>
INFO: Repository cuda_cusparse instantiated at:
  /home/johnny/Projects/jax/WORKSPACE:81:30: in <toplevel>
  /home/johnny/.cache/bazel/_bazel_johnny/ec3d842aa30ca8685e26a318bc6b64a6/external/tsl/third_party/gpus/cuda/hermetic/cuda_redist_init_repositories.bzl:538:18: in cuda_redist_init_repositories
Repository rule cuda_repo defined at:
  /home/johnny/.cache/bazel/_bazel_johnny/ec3d842aa30ca8685e26a318bc6b64a6/external/tsl/third_party/gpus/cuda/hermetic/cuda_redist_init_repositories.bzl:350:28: in <toplevel>
INFO: Repository cuda_nvtx instantiated at:
  /home/johnny/Projects/jax/WORKSPACE:81:30: in <toplevel>
  /home/johnny/.cache/bazel/_bazel_johnny/ec3d842aa30ca8685e26a318bc6b64a6/external/tsl/third_party/gpus/cuda/hermetic/cuda_redist_init_repositories.bzl:538:18: in cuda_redist_init_repositories
Repository rule cuda_repo defined at:
  /home/johnny/.cache/bazel/_bazel_johnny/ec3d842aa30ca8685e26a318bc6b64a6/external/tsl/third_party/gpus/cuda/hermetic/cuda_redist_init_repositories.bzl:350:28: in <toplevel>
INFO: Repository cuda_cccl instantiated at:
  /home/johnny/Projects/jax/WORKSPACE:81:30: in <toplevel>
  /home/johnny/.cache/bazel/_bazel_johnny/ec3d842aa30ca8685e26a318bc6b64a6/external/tsl/third_party/gpus/cuda/hermetic/cuda_redist_init_repositories.bzl:538:18: in cuda_redist_init_repositories
Repository rule cuda_repo defined at:
  /home/johnny/.cache/bazel/_bazel_johnny/ec3d842aa30ca8685e26a318bc6b64a6/external/tsl/third_party/gpus/cuda/hermetic/cuda_redist_init_repositories.bzl:350:28: in <toplevel>
ERROR: An error occurred during the fetch of repository 'cuda_nvjitlink':
   Traceback (most recent call last):
	File "/home/johnny/.cache/bazel/_bazel_johnny/ec3d842aa30ca8685e26a318bc6b64a6/external/tsl/third_party/gpus/cuda/hermetic/cuda_redist_init_repositories.bzl", line 348, column 44, in _cuda_repo_impl
		_use_downloaded_cuda_redistribution(repository_ctx)
	File "/home/johnny/.cache/bazel/_bazel_johnny/ec3d842aa30ca8685e26a318bc6b64a6/external/tsl/third_party/gpus/cuda/hermetic/cuda_redist_init_repositories.bzl", line 314, column 55, in _use_downloaded_cuda_redistribution
		arch_key = OS_ARCH_DICT[_get_platform_architecture(repository_ctx)]
	File "/home/johnny/.cache/bazel/_bazel_johnny/ec3d842aa30ca8685e26a318bc6b64a6/external/tsl/third_party/gpus/cuda/hermetic/cuda_redist_init_repositories.bzl", line 288, column 46, in _get_platform_architecture
		uname_result = repository_ctx.execute("uname", "-a").stdout
Error in execute: in call to execute(), parameter 'arguments' got value of type 'string', want 'sequence'
INFO: Repository cuda_cudart instantiated at:
  /home/johnny/Projects/jax/WORKSPACE:81:30: in <toplevel>
  /home/johnny/.cache/bazel/_bazel_johnny/ec3d842aa30ca8685e26a318bc6b64a6/external/tsl/third_party/gpus/cuda/hermetic/cuda_redist_init_repositories.bzl:538:18: in cuda_redist_init_repositories
Repository rule cuda_repo defined at:
  /home/johnny/.cache/bazel/_bazel_johnny/ec3d842aa30ca8685e26a318bc6b64a6/external/tsl/third_party/gpus/cuda/hermetic/cuda_redist_init_repositories.bzl:350:28: in <toplevel>
ERROR: /home/johnny/Projects/jax/WORKSPACE:81:30: fetching cuda_repo rule //external:cuda_nvjitlink: Traceback (most recent call last):
	File "/home/johnny/.cache/bazel/_bazel_johnny/ec3d842aa30ca8685e26a318bc6b64a6/external/tsl/third_party/gpus/cuda/hermetic/cuda_redist_init_repositories.bzl", line 348, column 44, in _cuda_repo_impl
		_use_downloaded_cuda_redistribution(repository_ctx)
	File "/home/johnny/.cache/bazel/_bazel_johnny/ec3d842aa30ca8685e26a318bc6b64a6/external/tsl/third_party/gpus/cuda/hermetic/cuda_redist_init_repositories.bzl", line 314, column 55, in _use_downloaded_cuda_redistribution
		arch_key = OS_ARCH_DICT[_get_platform_architecture(repository_ctx)]
	File "/home/johnny/.cache/bazel/_bazel_johnny/ec3d842aa30ca8685e26a318bc6b64a6/external/tsl/third_party/gpus/cuda/hermetic/cuda_redist_init_repositories.bzl", line 288, column 46, in _get_platform_architecture
		uname_result = repository_ctx.execute("uname", "-a").stdout
Error in execute: in call to execute(), parameter 'arguments' got value of type 'string', want 'sequence'
ERROR: @local_config_cuda//:enable_cuda :: Error loading option @local_config_cuda//:enable_cuda: no such package '@cuda_nvjitlink//': in call to execute(), parameter 'arguments' got value of type 'string', want 'sequence'
Traceback (most recent call last):
  File "/home/johnny/Projects/jax/build/build.py", line 721, in <module>
    main()
  File "/home/johnny/Projects/jax/build/build.py", line 672, in main
    shell(build_cpu_wheel_command)
  File "/home/johnny/Projects/jax/build/build.py", line 45, in shell
    output = subprocess.check_output(cmd)
  File "/usr/lib/python3.10/subprocess.py", line 421, in check_output
    return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
  File "/usr/lib/python3.10/subprocess.py", line 526, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['./bazel-6.5.0-linux-arm64', 'run', '--verbose_failures=true', '//jaxlib/tools:build_wheel', '--', '--output_path=/home/johnny/Projects/jax/dist', '--jaxlib_git_hash=b886bd7300376efc59fd9a159fa83314ed8c51a4', '--cpu=aarch64']' returned non-zero exit status 2.

```

Reply:
maybe:
uname_result = repository_ctx.execute(["uname", "-a"]).stdout or uname_result = repository_ctx.execute(["uname"], arguments=["-a"]).stdout ?

Reply:
Sorry, I've updated the PR. Please try again.

Reply:
Good news! Please let me know if downloading `linux-aarch64` redistributions helps you to build the artifacts on Jetson platform.

Reply:
omg! Now it is very easy to compile it! Thank you so much :)

Reply:
Glad I was able to help! I'll send out my PR for the internal review.

Reply:
https://github.com/openxla/xla/pull/17149 is submitted.